"3a Explain how k-fold cross-validation is implemented."

When we use the k-fold cross-validation, the dataset is splited in k-numbered different datasets, of approximately equal size. K-1 datasets are used to fit the method, which is then tested on the one remaining dataset (validation set). This whole procedure is repeated k-times, so that always new datasets will be drawn.

The k-fold cross-validation can be implemented with the cv.glm() function.


"3b What are the advantages and disadvantages of k-fold cross-validation relative to:" 

validation set approach: The k-fold cross-validation doesn't overestimate the test error rate, because it does not only fit the method on one half of the dataset (as with the validation set approach), but it uses a lot of different datasets on which the method is fitted on.


LOOCV: One adavantage is that the k-fold cross-validation is easier to compute since less code needs to be written. Another one is that k-fold Cross-validation often gives more accurate estimates of the test error rate than LOOCV. Furthermore LOOCV has a higher variance than the k-fold cross-validation with k< n.

In [34]:
"7a"
library(ISLR)
library (boot)

glm_a <- glm(Direction~Lag1+Lag2, data=Weekly, family=binomial)
summary(glm_a)

[1] "7a"


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.623  -1.261   1.001   1.083   1.506  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22122    0.06147   3.599 0.000319 ***
Lag1        -0.03872    0.02622  -1.477 0.139672    
Lag2         0.06025    0.02655   2.270 0.023232 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1488.2  on 1086  degrees of freedom
AIC: 1494.2

Number of Fisher Scoring iterations: 4


In [35]:
"7b"

glm_b <- glm(Direction~Lag1+Lag2, data=Weekly [-1,], family=binomial)
summary(glm_b)

[1] "7b"


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly[-1, 
    ])

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6258  -1.2617   0.9999   1.0819   1.5071  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22324    0.06150   3.630 0.000283 ***
Lag1        -0.03843    0.02622  -1.466 0.142683    
Lag2         0.06085    0.02656   2.291 0.021971 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1494.6  on 1087  degrees of freedom
Residual deviance: 1486.5  on 1085  degrees of freedom
AIC: 1492.5

Number of Fisher Scoring iterations: 4


In [36]:
"7c"
predict.glm(glm_b, Weekly[1, ], type = "response") > 0.5
head(Weekly)

[1] "7c"

1 
TRUE

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1 1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2 1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3 1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4 1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5 1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6 1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down

No, it was not correctly classified. The prediction was "Up", but the correct "answer" would be "Down".

In [37]:
"7d"
count = rep(0, dim(Weekly)[1])
for (i in 1:(dim(Weekly)[1])) {
#i
glm_d <- glm(Direction~Lag1+Lag2, data=Weekly [-i,], family=binomial) 
#ii
is_up <- predict.glm(glm_d, Weekly[i, ], type = "response") > 0.5
#iii
is_true_up = Weekly[i, ]$Direction == "Up"
    if (is_up != is_true_up) 
        count[i] = 1
}
#iv
sum(count)

[1] "7d"

[1] 490

490 mistakes were made.

In [38]:
"7e"
mean(count)

[1] "7e"

[1] 0.4499541

LOOCV estimates a test error rate of approximately 45%. 

In [39]:
"9"
library(MASS)
summary(Boston)

[1] "9"

      crim                zn             indus            chas        
 Min.   : 0.00632   Min.   :  0.00   Min.   : 0.46   Min.   :0.00000  
 1st Qu.: 0.08205   1st Qu.:  0.00   1st Qu.: 5.19   1st Qu.:0.00000  
 Median : 0.25651   Median :  0.00   Median : 9.69   Median :0.00000  
 Mean   : 3.61352   Mean   : 11.36   Mean   :11.14   Mean   :0.06917  
 3rd Qu.: 3.67708   3rd Qu.: 12.50   3rd Qu.:18.10   3rd Qu.:0.00000  
 Max.   :88.97620   Max.   :100.00   Max.   :27.74   Max.   :1.00000  
      nox               rm             age              dis        
 Min.   :0.3850   Min.   :3.561   Min.   :  2.90   Min.   : 1.130  
 1st Qu.:0.4490   1st Qu.:5.886   1st Qu.: 45.02   1st Qu.: 2.100  
 Median :0.5380   Median :6.208   Median : 77.50   Median : 3.207  
 Mean   :0.5547   Mean   :6.285   Mean   : 68.57   Mean   : 3.795  
 3rd Qu.:0.6240   3rd Qu.:6.623   3rd Qu.: 94.08   3rd Qu.: 5.188  
 Max.   :0.8710   Max.   :8.780   Max.   :100.00   Max.   :12.127  
      rad              tax 

In [40]:
"9a"
μ = mean(Boston$medv)
μ

[1] "9a"

[1] 22.53281

In [41]:
"9b"
μ.err = sd(Boston$medv)/sqrt(length(Boston$medv))
μ.err

[1] "9b"

[1] 0.4088611

In [42]:
"9c"
install.packages("boot")
library(boot)
boot_c = function(data, index) return(mean(data[index]))
library(boot)
bstrap = boot(Boston$medv, boot_c, 1000)
bstrap

[1] "9c"

Warning message:
"package 'boot' is in use and will not be installed"



ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Boston$medv, statistic = boot_c, R = 1000)


Bootstrap Statistics :
    original      bias    std. error
t1* 22.53281 0.005967391   0.4172546

The two standard error rates are quiet similar 0.409 vs. 0.399.

In [43]:
"9d"
t.test(Boston$medv)


[1] "9d"


	One Sample t-test

data:  Boston$medv
t = 55.111, df = 505, p-value < 2.2e-16
alternative hypothesis: true mean is not equal to 0
95 percent confidence interval:
 21.72953 23.33608
sample estimates:
mean of x 
 22.53281 


In [44]:
c(bstrap$t0 - 2 * 0.3993, bstrap$t0 + 2 * 0.3993)

[1] 21.73421 23.33141

The result of the t-test lies within the bootstrap estimated confident interval. The mean of x from the t-test +/-0.798 results in the two borders of the confident intervals

In [45]:
"9e"
μ_med = median(Boston$medv)
μ_med

[1] "9e"

[1] 21.2

In [46]:
"9f"
boot_f <- function(data, index) return(median(data[index]))
boot(Boston$medv, boot_f, 1000)

[1] "9f"


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Boston$medv, statistic = boot_f, R = 1000)


Bootstrap Statistics :
    original   bias    std. error
t1*     21.2 -0.02555   0.3729327

The median is 21.2 with standard error of 0.38, which is relatively small.

In [47]:
"9g"
μ0.1 <- quantile(Boston$medv, 0.1)
μ0.1

[1] "9g"

10% 
12.75

In [48]:
"9h"
boot_h <- function(data, index) return(quantile(data[index], c(0.1)))
boot(Boston$medv, boot_h, 1000)

[1] "9h"


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Boston$medv, statistic = boot_h, R = 1000)


Bootstrap Statistics :
    original  bias    std. error
t1*    12.75 0.01365   0.4947787

The tenth-percentile value is 12.75 with a standard error of 0.5, which is relatively.